timesketch/ui/graphs/graph-view.component.ts

```diff
@@ -0,0 +1,130 @@
+import {Component, Input} from '@angular/core'
+
+import {CytoscapeSettings} from './cytoscape-settings.component'
+
+export type GraphViewState = {
```

Comment:
I wonder if it wouldn't just be easier to write

export interface GraphViewState {
  type: 'empty'|'loading'|'ready';
  element?: ...
}

Reply:
But this will no longer be semantically correct, right? I imagine later on this could become something like:
```typescript
type GraphViewState = {
    type: 'empty' | 'loading' | 'ready' | 'error',
    spinnerUrl?: string,
    message?: string,
    elements?: Cy.ElementsDefinition,
    elementTypes?: ElementType[],
}
```
How do you figure out which fields are available for which type?

Reply:
Related: http://blog.jenkster.com/2016/06/how-elm-slays-a-ui-antipattern.html

Reply:
This would not be semantically identical to what is written now. It's ok to keep the discriminator columns and be more precise on types, but in such a scenario, I would split this single type definition into multiple definitions, e.g.

```
export type Empty = { ... };
export type Loading = {...};
export type GraphViewState = Empty | Loading | ... ;
```

Reply:
Done.